src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java

```diff
@@ -1085,47 +1102,139 @@ private static final class SimpleStringBuilderStrategy {
                     new Consumer<ClassBuilder>() {
                         @Override
                         public void accept(ClassBuilder clb) {
                             clb.withFlags(ClassFile.ACC_FINAL | ClassFile.ACC_SUPER | ClassFile.ACC_SYNTHETIC)
+                                .withField(LENGTH, CD_int, FINAL_STABLE_FIELD)
+                                .withField(CODER, CD_byte, FINAL_STABLE_FIELD)
+                                .withField(CONSTANTS, CD_Array_String, FINAL_STABLE_FIELD)
+                                .withMethodBody("<init>",
+                                        ARRAY_STRING_TO_VOID,
+                                        ClassFile.ACC_PRIVATE,
+                                        generateConstructor(concatClass, constants.length)
+                                )
                                 .withMethodBody(METHOD_NAME,
-                                        ConstantUtils.methodTypeDesc(erasedArgs),
-                                        ClassFile.ACC_FINAL | ClassFile.ACC_PRIVATE | ClassFile.ACC_STATIC,
-                                        generateMethod(constants, erasedArgs));
+                                        ConstantUtils.methodTypeDesc(concatArgs),
+                                        ClassFile.ACC_FINAL | ClassFile.ACC_PRIVATE,
+                                        generateConcatMethod(concatClass, concatArgs));
                     }});
             try {
                 var hiddenClass = lookup.makeHiddenClassDefiner(className, classBytes, Set.of(), DUMPER)
                         .defineClass(true, null);
-                return lookup.findStatic(hiddenClass, METHOD_NAME, erasedArgs);
-            } catch (Exception e) {
+
+                MethodHandle constructorHandle = lookup.findConstructor(hiddenClass, MethodType.methodType(void.class, String[].class));
+                var instance = hiddenClass.cast(constructorHandle.invoke(constants));
+                MethodHandle handle = lookup.findVirtual(hiddenClass, METHOD_NAME, concatArgs);
+                CACHE.put(concatArgs, new SoftReference<>(new MethodHandlePair(constructorHandle, handle)));
+                return handle.bindTo(instance);
+            } catch (Throwable e) {
                 throw new StringConcatException("Exception while spinning the class", e);
             }
         }
 
+        private static Consumer<CodeBuilder> generateConstructor(ClassDesc concatClass, int numConstants) {
+            return new Consumer<CodeBuilder>() {
+                @Override
+                public void accept(CodeBuilder cb) {
+                    cb.aload(0);
+                    cb.invokespecial(CD_Object, INIT_NAME, MTD_void, false);
+                    cb.aload(0);
+                    cb.aload(1);
+                    cb.putfield(concatClass, CONSTANTS, CD_Array_String);
+                    cb.iconst_0();
+                    cb.istore(2);
```

Comment:
These are too confusing. Maybe use `int lengthSlot = cb.allocateLocal(TypeKind.IntType);` instead, similar for other local vars.